docs/site/Binding.md

```diff
@@ -83,6 +83,21 @@ binding.toDynamicValue(() => new Date());
 binding.toDynamicValue(() => Promise.resolve('my-value'));
 ```
 
+The factory function can receive extra information about the context, binding,
+and resolution options.
+
+```ts
+import {ValueFactory} from '@loopback/context';
+
+// The factory function now have access extra metadata about the resolution
+const factory: ValueFactory<string> = (_ctx, _binding, options) => {
+  return `Hello, ${_ctx.name}#${
+    _binding.key
+  } ${options.session?.getBindingPath()}`;
+};
+const b = ctx.bind('msg').toDynamicValue(factory);
+```
+
```

Comment:
In the past, I was against allowing dynamic value factories to access the context object, because it makes it too easy to slip into using Service Locator design pattern (`ctx.get('user')`) instead of the preferred Dependency Injection pattern.

Considering how little was done for dynamic values in the last three years, I agree it's a good idea to allow them to access the context (and other metadata) directly :+1:

At the same time, I would like to keep the doors open to allow us to implement dependency injection for dynamic values too, e.g.

```ts
ctx.bind('msg').toDynamicValue(SecurityBindings.USER, (ctx, user) => {
  return `Hello ${user.name}`;
});

// Note that `user` is strongly typed as `UserProfile`,
// using the type info provided by USER binding key
```

I feel the current design, where you are adding three arguments (`ctx`, `binding`, `options`), could make it difficult to come up with an ergonomic API for injecting values to dynamic value factories in the future.

I am proposing to convert these arguments into a single object argument.

```ts
const factory: ValueFactory<string> = ({context, binding, options}) => {
   return `Hello, ${context.name}#${
     binding.key
   } ${options.session?.getBindingPath()}`;
 };
```

I am surprised we don't have a type to hold all this fields yet, e.g. `ResolutionContext`, conceptually similar to `InvocationContext` and `RequestContext`.

I also noticed that `ResolverFunction` has similar but slightly different signature:

https://github.com/strongloop/loopback-next/blob/fcace9ea07c2d218835a55f3bc2c249ce0963620/packages/context/src/inject.ts#L46-L51

The good part is that the session can be accessed from dynamic value factories via `ResolutionOptions.session` :+1: 

The bad part is that we have different flavors of resolvers:
- Dynamic value factory receives full resolution options
- Resolvers receive the current session only

Have you considered any alternative designs that would be more consistent? Maybe this inconsistency is just a price we have to pay for evolving design and we should put a TODO item to modify `ResolveFactory` to accept full resolution options when we are working on the next major release of `@loopback/context`?

As I am thinking about this some more, I guess it would be best to:
- Keep `ResolverFunction` as is to preserve backwards compatibility.
- Implement `ValueFactory` using the design we want to use going forward (i.e. with access to full options).
  - Personally, I strongly prefer `ResolutionContext` over positional arguments. Besides the arguments posted above, this approach is also more future proof.
- Add `TODO(semver-major)` to update `ResolverFunction` to the new style in the future.

Reply:
I added some comments to `ResolverFn` and defer the refactoring to the future.